### ISSUE TITLE
MSMQ with Native transaction deadlocks when the MSMQ service is stopped or restarted

### ISSUE THREAD

**Issue by danielmarbach** [Particular/NServiceBus#4188]
## Who's affected

* MSMQ users using native transactions (`busConfiguration.Transactions().DisableDistributedTransactions();`) with NServiceBus v5

## Symptoms

* When a receiving endpoint is using native MSMQ transactions and the MSMQ service restarts while the endpoint is running then the endpoint deadlocks and cannot be properly shutdown. 

## Steps to reproduce

* Use full duplex sample, change the server bus configuration to `busConfiguration.Transactions().DisableDistributedTransactions();` send a few messages from the client to the server to warm it up
* Stop the MSMQ service

The following log statement will be visible

```
2016-10-05 11:25:16.840 INFO  NServiceBus.Unicast.Transport.TransportReceiver Failed to process message
System.Messaging.MessageQueueException (0x80004005): Message Queue service is not available.
   at System.Messaging.MessageQueueTransaction.Begin()
   at NServiceBus.Transports.Msmq.MsmqDequeueStrategy.Action() in C:\Build\src\NServiceBus.Core\Transports\Msmq\MsmqDequeueStrategy.cs:line 201
2016-10-05 11:25:16.841 INFO  NServiceBus.Unicast.Transport.TransportReceiver Failed to process message
System.Messaging.MessageQueueException (0x80004005): Message Queue service is not available.
   at System.Messaging.MessageQueueTransaction.Begin()
   at NServiceBus.Transports.Msmq.MsmqDequeueStrategy.Action() in C:\Build\src\NServiceBus.Core\Transports\Msmq\MsmqDequeueStrategy.cs:line 201
2016-10-05 11:25:16.841 INFO  NServiceBus.Unicast.Transport.TransportReceiver Failed to process message
System.Messaging.MessageQueueException (0x80004005): Message Queue service is not available.
   at System.Messaging.MessageQueueTransaction.Begin()
   at NServiceBus.Transports.Msmq.MsmqDequeueStrategy.Action() in C:\Build\src\NServiceBus.Core\Transports\Msmq\MsmqDequeueStrategy.cs:line 201
2016-10-05 11:25:16.841 INFO  NServiceBus.Unicast.Transport.TransportReceiver Failed to process message
System.Messaging.MessageQueueException (0x80004005): Message Queue service is not available.
   at System.Messaging.MessageQueueTransaction.Begin()
   at NServiceBus.Transports.Msmq.MsmqDequeueStrategy.Action() in C:\Build\src\NServiceBus.Core\Transports\Msmq\MsmqDequeueStrategy.cs:line 201
```

The endpoint never recovers from that and cannot be shutdown. It hangs forever on

`2016-10-05 11:27:48.976 INFO  NServiceBus.Unicast.UnicastBus Initiating shutdown.`

### Analysis

Originally raised by @johnsimons in https://github.com/Particular/NServiceBus/issues/4109

Annotations as comments inline in the code

```
void OnPeekCompleted(object sender, PeekCompletedEventArgs peekCompletedEventArgs)
        {
            // behaves properly even when service is shutdown
            CallPeekWithExceptionHandling(() => queue.EndPeek(peekCompletedEventArgs.AsyncResult));

            ... omitted for clarity

            // Problem is caused inside action with native tx
           Task.Factory
                .StartNew(Action, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default)...

            // DEADLOCK HERE
            peekResetEvent.WaitOne();

            // if we manage to reach this line the PeekCompleted will be raised even when the service is shut down so we are good
            CallPeekWithExceptionHandling(() => queue.BeginPeek());

           ... omitted for clarity
        }
```
Offending code
``` 
void Action()
        {
            TransportMessage transportMessage = null;
            try
            {
                if (transactionSettings.IsTransactional)
                {
                    if (transactionSettings.SuppressDistributedTransactions)
                    {
                        using (var msmqTransaction = new MessageQueueTransaction())
                        {
                            // When this throws when the service is offline we will never set the peekResetEvent and therefore deadlock on the peekResetEvent.WaitOne() call
                            msmqTransaction.Begin();
                            
                            Message message;
                            // This call would release the peekResetEvent if we were able to reach it but we aren't when the service is offline
                            if (!TryReceiveMessage(() => queue.Receive(receiveTimeout, msmqTransaction),out message))
         ... omitted for clarity
``` 

### Potential Fix

Move the `msmqTransaction.Begin();` into the lambda expression of `TryReceive` like

```
!TryReceiveMessage(() => {
   msmqTransaction.Begin();
   return queue.Receive(receiveTimeout, msmqTransaction);
})
```

Remove the transaction commit in the if statement (need to double check)

We will backport this to 5.2.18. Potentially we should Upgrade ServiceControl to that version since SC uses native transactions and therefore is subjected to this dead lock // cc @Particular/servicecontrol-maintainers 

### Minimal repro to highlight the problem

```
using System;
using System.Linq;
using System.Messaging;
using System.ServiceProcess;
using System.Threading;
using System.Threading.Tasks;
using System.Transactions;

namespace MsmqRepro
{
    class Program
    {
        private static MessageQueue receiveQueue;
        private static int numberOfMessages = 0;

        static void Main(string[] args)
        {
            selectedMode = args.Select(s => Enum.Parse(typeof(Mode), s)).OfType<Mode>().SingleOrDefault();
            Console.WriteLine("Mode: "+  selectedMode);

            StartMsmq();

            receiveQueue = new MessageQueue(".\\private$\\reproqueue", false, false, QueueAccessMode.Receive);
            var sendQueue = new MessageQueue(".\\private$\\reproqueue", false, true, QueueAccessMode.Send);

            var messageReadPropertyFilter = new MessagePropertyFilter
            {
                Body = true,
                TimeToBeReceived = true,
                Recoverable = true,
                Id = true,
                ResponseQueue = true,
                CorrelationId = true,
                Extension = true,
                AppSpecific = true
            };

            receiveQueue.MessageReadPropertyFilter = messageReadPropertyFilter;

            var tx = new MessageQueueTransaction();
            tx.Begin();
            for (int i = 0; i < 20; i++)
            {
                sendQueue.Send(new Message {Label = "TestMessage" + i}, tx);
            }
            tx.Commit();

            MessageQueue.ClearConnectionCache();

            receiveQueue.PeekCompleted += Queue_PeekCompleted;
            receiveQueue.BeginPeek();

            Console.ReadLine();

            StartMsmq();
        }

        private static void Queue_PeekCompleted(object sender, PeekCompletedEventArgs e)
        {
            CallPeekWithExceptionHandling(() => receiveQueue.EndPeek(e.AsyncResult));

            Console.Write(".");
            Thread.Sleep(100);

            numberOfMessages++;
            if (numberOfMessages == 10)
                StopMsmq();

            Task.Factory
                .StartNew(Action, CancellationToken.None, TaskCreationOptions.LongRunning, TaskScheduler.Default)
                .ContinueWith(task => task.Exception.Handle(ex =>
                {
                    Console.WriteLine("Error processing message.");
                    return true;
                }), TaskContinuationOptions.OnlyOnFaulted);

            // Deadlock
            peekResetEvent.WaitOne();

            CallPeekWithExceptionHandling(() => receiveQueue.BeginPeek());
        }

        static AutoResetEvent peekResetEvent = new AutoResetEvent(false);
        private static Mode selectedMode;


        static void Action()
        {
            Message message;
            switch (selectedMode)
            {
                case Mode.Native:
                    using (var transaction = new MessageQueueTransaction())
                    {
                        transaction.Begin();

                        if (
                            !TryReceiveMessage(() => receiveQueue.Receive(TimeSpan.FromSeconds(1), transaction),
                                out message))
                        {
                            transaction.Commit();
                            return;
                        }

                        transaction.Commit();
                    }
                    break;
                case Mode.Dtc:
                    using (var scope = new TransactionScope(TransactionScopeOption.Required))
                    {
                        if (!TryReceiveMessage(() => receiveQueue.Receive(TimeSpan.FromSeconds(1), MessageQueueTransactionType.Automatic), out message))
                        {
                            scope.Complete();
                            return;
                        }

                        // Do Something
                        scope.Complete();
                    }
                    break;
                case Mode.None:

                    if (!TryReceiveMessage(() => receiveQueue.Receive(TimeSpan.FromSeconds(1), MessageQueueTransactionType.None), out message))
                    {
                        return;
                    }
                    break;
            }
        }

        static bool TryReceiveMessage(Func<Message> receive, out Message message)
        {
            message = null;

            try
            {
                message = receive();
                return true;
            }
            catch (MessageQueueException messageQueueException)
            {
                if (messageQueueException.MessageQueueErrorCode == MessageQueueErrorCode.IOTimeout)
                {
                    //We should only get an IOTimeout exception here if another process removed the message between us peeking and now.
                    return false;
                }

                Console.WriteLine("Critical");
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error in receiving messages.", ex);
            }
            finally
            {
                peekResetEvent.Set();
            }

            return false;
        }

        private static void StopMsmq()
        {
            Console.Write("!!!");
            ServiceController service = new ServiceController("MSMQ");
            try
            {
                TimeSpan timeout = TimeSpan.FromMilliseconds(1000);

                service.Stop();
                service.WaitForStatus(ServiceControllerStatus.Stopped, timeout);
            }
            catch
            {
                // ...
            }
        }

        private static void StartMsmq()
        {
            ServiceController service = new ServiceController("MSMQ");
            try
            {
                TimeSpan timeout = TimeSpan.FromMilliseconds(1000);

                service.Start();
                service.WaitForStatus(ServiceControllerStatus.Running, timeout);
            }
            catch
            {
                // ...
            }
        }

        static void CallPeekWithExceptionHandling(Action action)
        {
            try
            {
                action();
            }
            catch (MessageQueueException messageQueueException)
            {
                Console.Write("e");
            }
        }
    }

    enum Mode
    {
        Native,
        Dtc, 
        None
    }
}

```

#### Native

```
Mode: Native
..........!!!Error processing message.
```
No more messages are processed. The process deadlocks

#### Dtc

```
Mode: Dtc
..........!!!Critical
e.Critical
e.Critical
e.Critical
...
```

Properly raises critical

#### None

```
Mode: None
..........!!!Critical
e.Critical
e.Critical
e.Critical
...
```
Properly raises critical

**Comment by johnsimons** [comment]
Yes we need this patch for ServiceControl

**Comment by timbussmann** [comment]
will be released as hotfix 5.2.18
